### ISSUE TITLE
OutOfMemoryError when number of mesh vertices > 32767

### ISSUE THREAD

**Issue by aliasaila** [GoranM/bdx#537]
libgdx doesn't support more than 32767 vertices within a single mesh, so there needs to be some additional handling for when this limit is exceeded.

[https://github.com/libgdx/libgdx/wiki/Importing-Blender-models-in-LibGDX#maximum-vertices](https://github.com/libgdx/libgdx/wiki/Importing-Blender-models-in-LibGDX#maximum-vertices)

**Comment by Fr0sk** [comment]
Maybe we can throw a pop-up warning explaining the situation and asking the developer to either use a decimate modifier, use a [mesh poly reducer script](http://wiki.blender.org/index.php/Extensions:2.4/Py/Scripts/Mesh/Mesh_poly_reduce) (both of these automatic) or manually strip down the mesh in several vertex groups. I don't recall if there is a way of doing the last one automatically, but maybe that would result in unwanted behavior. But this limit is not libgdx or bdx specific. Most engines have one of those limits specified on the link you posted, so maybe a warning would be just enough. Just my 2 cents though.

**Comment by SolarLune** [comment]
Hm... Since the `short`'s limit only influences the number of vertices in the mesh when we're actually building the mesh, it might be possible to auto-split the mesh according to the number of remaining vertices to build (i.e. it creates one MeshPart per 32767 vertices per material, rather than just creating one MeshPart per material). 

This would mess with the material switching, though, as each MeshPart is currently treated as a separate material slot. This could be solved by simply giving each MeshPart name its material slot number. Then, when we want to change the material in a slot, we can loop through the MeshParts and change the material of the parts that have that material ID.

So basically, can we just automatically split up meshes that exceed 32767 vertices?

**Comment by Fr0sk** [comment]
That would not break moving meshes and possibly bone animated meshes too?

**Comment by SolarLune** [comment]
We don't have bone animated meshes, so I dunno about that. So maybe we should wait until we have animated bone support, I guess? To make sure it doesn't break it?

**Comment by Fr0sk** [comment]
Since it's on a issue, I though it was somewhat on the way. If we create a empty holder object and then subdivide the mesh and parent each part to the holder, and the armature too, that would work?

**Comment by SolarLune** [comment]
> Since it's on a issue, I though it was somewhat on the way.

Basically, it being on an issue just means it's something that someone can tackle next, not that it's necessarily going to be finished soon. It looks kinda complex, to me; I created the issue to basically ask to see what the process was and to get the ball rolling, so to speak. I haven't personally done any work on armatures other than just info-gathering.

> If we create a empty holder object and then subdivide the mesh and parent each part to the holder, and the armature too, that would work? I mean, in a normal blender project, the animations would still work (I'm not used to work with bone animations)?

Not sure what you mean here. Do you mean create an empty holder object in Blender, or in BDX, or what?

> Is it possible to only split on the build process and join again in a binary exported file, or something like that?

The split would only be necessary when building the mesh in BDX, as LibGDX uses shorts for the vertex index. So basically, it's all one mesh from Blender, and in the exported file. We'd only need to "do something" when we actually go to reconstruct the mesh from the exported data.

**Comment by GoranM** [comment]
This is something that should be fixed in libgdx.

If there's really a need for a workaround in BDX, then it's something worth considering, but there has to a non-contrived example that makes that clear.

**Comment by Fr0sk** [comment]
>If there's really a need for a workaround in BDX, then it's something worth considering, but there has to a non-contrived example that makes that clear.

The only I could think of is a full terrain with hills. If the terrain is big enough and not much flat, it will need many vertices to keep the detail. OpenGL and therefore LibGDX support unsigned indexes, so it would allow double the amount of vertices, but as they say, may result in unexpected results in some devices. I think that throwing an exception saying to reduce the vertex count would be enough, it gives the developer the freedom of choice to decide what's better. As I said, the 32767 is not exclusive to LibGDX.

**Comment by aliasaila** [comment]
> I think that throwing an exception saying to reduce the vertex count would be enough

I agree.

The libgdx wiki gives fair warning about this limitation and states that it's the responsibility of the user to "make sure that your meshes never exceed this limitation".

**Comment by GoranM** [comment]
Throwing an exception makes sense.

**Comment by Fr0sk** [comment]
@aliasaila Do you want to make the code changes or don't mind if I do?

**Comment by aliasaila** [comment]
@Fr0sk If you have any ideas, go for it.

**Comment by SolarLune** [comment]
About @Fr0sk's PR:

I think it best that the exception be thrown in the exporter rather than the engine, as it's quickest to simply throw the exception when you try to run BDX. The limitation may be on the LibGDX side, but the only way to resolve the problem is to reduce the vertex count in Blender. So, it makes sense that the exception appear in Blender.

**Comment by SolarLune** [comment]
> This is something that should be fixed in libgdx.

Coming back to this, I agree that it would be nice to have more vertices, but I think it might still be best to try a solution on our end. Like @Fr0sk mentioned, I'm trying to do a simple test terrain. With 128 subdivisions (19,288 vertices), BDX crashes due to an OutOfMemoryError.

[There is an existing LibGDX tracker issue about this](https://github.com/libgdx/libgdx/issues/3386), and it was closed. The solution seems to be to split up meshes that are too big to lessen draw calls and to be culled by the frustum.

So, we have two solutions: 

- We could try splitting complex meshes up in Blender automatically before exporting them, and informing people in the console / through our documentation.
- We could try splitting meshes up in BDX when building them.

We could also, apparently, use unsigned shorts when building the meshes to raise the limit to double it, but apparently that's not suggested, according to that issue.

**Comment by Fr0sk** [comment]
> With 128 subdivisions (19,288 vertices), BDX crashes due to an OutOfMemoryError.

19,288 should still be working allright :s 

Either way, we have to split the mesh somewhere, and since faces lose connection I think it might break animations. Only testing one can be certain though

**Comment by Fr0sk** [comment]
I read somewhere that Unity vertex limit per mesh is 64k and unreal was the same, until recently unreal bumped that up. Still, if we manage to get the unsigned short working across all platforms it would be useful, at least to be in the same level as probably most engines out there

**Comment by SolarLune** [comment]
Yeah, I'm definitely getting an OOM error on less than 32,768 vertices.  I'm gonna guess that if we're duplicating vertices somehow.

> Either way, we have to split the mesh somewhere, and since faces lose connection I think it might break animations. Only testing one can be certain though

Well, the options are to wait until animations are implemented, or just do it and worry about animations later, when they're implemented.

> Still, if we manage to get the unsigned short working across all platforms it would be useful, at least to be in the same level as probably most engines out there

I don't know if that's something we should try - like I said, the issue I linked to mentioned that this wasn't "practical"? Maybe it's something that would have to be done in LibGDX; not sure.

**Comment by Fr0sk** [comment]
> Well, the options are to wait until animations are implemented, or just do it and worry about animations later, when they're implemented.

I think that animations aren't coming that soon, so yeah, if they do eventually break, lets worry about them later.

> I don't know if that's something we should try - like I said, the issue I linked to mentioned that this wasn't "practical"? Maybe it's something that would have to be done in LibGDX; not sure.

Probably is, but they do not seem to be interested in implementing that. At least not for now. Games made with LibGDX tend to be 2D or really simple 3D games, so I think that 32k cap isn't that much of a problem. Since BDX is a 3D engine, it probably will be using meshes with higher poly count than most LibGDX games. If that's a way to go, if it might be easyer to implement and still be fully functional, I think we should at least considering doing it, even if it isn't supported by default on LibGDX.

**Comment by SolarLune** [comment]
OK, we must have duplicates, here - I have a human model that's just 8,782 vertices / 17,539 faces, and it's crashing. BDX is reporting 51240 vertices, which is, of course, waaaaay too much, and even that's with the dude's head lopped off, haha. 

Gonna look into this.

**Comment by Fr0sk** [comment]
> OK, we must have duplicates, here - I have a human model that's just 8,782 vertices / 17,539 faces, and it's crashing. BDX is reporting 51240 vertices

A single mesh or combined meshes? Because that seems an odd number, since they are not multiple :s good luck, and don't lose your head like your dude :D

**Comment by SolarLune** [comment]
> A single mesh or combined meshes?

A mesh with two materials, so I think this is the greater portion of the mesh.

> don't lose your head like your dude :D

Haha, I'll try not to.

So it looks like it is making duplicate vertices - one vertex for each polygon exported (so, for a cube, it's exporting 36 vertices, rather than just 8). I _think_ a solution would be to export just the vertices, as well as a triangle > vertex list.

**Comment by aliasaila** [comment]
@SolarLune Some vertices will share the same position but have different normals, and in those cases the seeming duplication is necessary.

If we build the index array in the exporter, and export only unique vertices, we can reduce the number of vertices by a third (for cubes and uv spheres anyway, ico spheres are unaffected).

In preliminary testing (with vsync disabled) it seems that this increases FPS by right around 30 percent.

I think it's worth doing.

**Comment by SolarLune** [comment]
Yeah, I think I agree, especially considering that the more vertices a model has, the closer it is to that short limit.

**Comment by GoranM** [comment]
> Some vertices will share the same position but have different normals, and in those cases the seeming duplication is necessary.

Similarly with UVs (texture coordinates).
